plasmapy/diagnostics/thomson.py

```diff
@@ -319,3 +345,207 @@ def spectral_density(
+):
+    r"""
+    Function that fits Thomson scattering spectra.
+    """
+
+    # Condition the raw input in the same way as the thomson.spectral_density
+    # function does
+    (
+        Te,
+        Ti,
+        efract,
+        ifract,
+        ion_species,
+        electron_vel,
+        ion_vel,
+    ) = _condition_thomson_input(
+        Te, Ti, efract, ifract, ion_species, electron_vel, ion_vel
+    )
+    # Extract the number of ion and electron populations
+    num_i = len(ifract)
+    num_e = len(efract)
+
+    settings = {
+        "ion_species": ion_species,
+        "probe_wavelength": probe_wavelength,
+        "probe_vec": probe_vec,
+        "scatter_vec": scatter_vec,
+        "num_i": num_i,
+        "num_e": num_e,
+    }
+
+    params = lmfit.Parameters()
+    params.add("n", value=n.to(u.cm ** -3).value, vary=False, min=0)
+
+    # Break apart vectorized inputs into individual quantities
+    # This is necessary because lmfit doesn't support vector parameters
+    for i in range(num_e):
+        params.add("efract" + str(i), value=efract[i], vary=False, min=0, max=1)
+        params.add("Te" + str(i), value=Te[i].to(u.K).value, vary=False, min=0)
+        for j in range(3):
+            params.add(
+                f"ve{i}_{j}", value=electron_vel[i, j].to(u.m / u.s).value, vary=False
+            )
+
+    for i in range(num_i):
+        params.add("ifract" + str(i), value=ifract[i], vary=False, min=0, max=1)
+        params.add("Ti" + str(i), value=Ti[i].to(u.K).value, vary=False, min=0)
+        for j in range(3):
+            params.add(
+                f"vi{i}_{j}", value=ion_vel[i, j].to(u.m / u.s).value, vary=False
+            )
+
+    # Catch references to single varaibles and append a 0
+    for v in fit:
+        if v in ["efract", "Te"] and num_e == 1:
+            fit.append(v + "0")
+            fit.remove(v)
+        elif v in ["ifract", "Ti"] and num_i == 1:
+            fit.append(v + "0")
+            fit.remove(v)
+
+    for v in fit:
+        valid_keys = list(params.keys())
+        if v in valid_keys:
+            params[v].vary = True
+        else:
+            raise KeyError(
+                f"Fit variable {v} is not valid for the provided "
+                f"fit parameters: {valid_keys}"
+            )
+
+    # params.pretty_print()
+
+    fit_fcn = lambda params, wavelengths, Skw: _spectral_density_fit_function(
+        params, wavelengths, Skw, settings
+    )
+
+    minner = lmfit.Minimizer(
+        fit_fcn,
+        params,
+        fcn_args=(wavelengths.to(u.nm).value, spectral_density.to(u.s / u.rad).value),
+    )
+
+    result = minner.minimize()
+
+    # lmfit.report_fit(result)
+
+    # TODO: if 0's were added (num_e or num_i==0), strip them off before reporting
+    # the best fit coefficents?
+
+    # TODO: What other fit information (chisqr, iter, etc.) should be included in
+    # the output?
+
+    best_fit = {}
+    for v in fit:
+        best_fit[v] = result.params[v].value
```

Comment:
Despite my comments above, I'm pretty sure we should output the error bars for the fit, anyway!

https://docs.astropy.org/en/stable/uncertainty/ might be good to take a look at, though I suspect the principle is a bit different: lmfit seems to output the point estimate to the parameter's value and its "standard error" (thus, assuming a Gaussian distribution of the parameter and taking the Gaussian's std, as I suspected); while `astropy.uncertainty` operates on the much simpler "here's the set of samples of my measurements, do the usual thing with no strings attached to get its mean/std/variance/other estimators".

Anyway, it'd be good to try to integrate this with as much existing functionality as possible, so we have to write as little custom code as necessary. Could be a future PR, though.

@rocco8773 I think you might want to take a look here?

Reply:
Absolutely agree - I'll be looking into this. I think I'll take the approach of directly exposing more of lmfits inputs and outputs to the user (eg. take Parameters as inputs, output a lmfit datastructure of some kind). I think if I demonstrate how to use them in the notebook that shouldn't be too confusing for new users?

Reply:
That sounds good! Sorry I missed responding here!